### DISCUSSION TITLE
Gitlab support

### DISCUSSION THREAD

**Discussion by jbpin** [tinacms/tinacms#2685] [Ideas]
It could be good to support Gitlab and leverage gitlab CI and Pages. 
As stated in the doc 
```
Tina's GraphQL Content API is flexible, in that it can be run locally using the Tina CLI ("Local Mode"), or your site can talk to our hosted content API in a production environment ("Prod Mode"), which persists changes to your GitHub repository.
```
What prevent you from supporting Gitlab out-of-the-box? 
I also found in the FAQ : 
> GitHub - The first Git provider that Tina Cloud integrates with. Other Git providers may be available in the future.

**Comment by scottgallant** [comment]
Thanks for posting, @jbpin.  I can say that we're _thinking_ about this, but have no concrete plans to support GitLab right now.   Do you use GitLab? Is this a blocker for you?

**Comment by suntudo** [comment]
Gitlab support would be huge for me and my team. We use Gitlab + Netlify for most things. Adopting Tina without Gitlab support is doable, but a big pain. Thanks for building this, btw.

**Reply by scottgallant** [reply]
We've been talking about this recently. I'd love to hear about your use-case to understand what types of teams/projects this is important for. Can you tell me about your use case at **scott at tina.io** ?

**Comment by jipiboily** [comment]


**Reply by jeffsee55** [reply]
> What does adding GitLab support involve? (curious to see why it's tied to one git host more than another)

Currently we're communicating directly to Github via its REST API, we're in the planning process of moving to a true Git-backed system which will be agnostic. At that point it won't matter where your content comes from. Sorry to say but we don't have a timeline for that yet.

> Does the free tier of GitHub can work? What are the limits, if any?

Yes you can use a free Github account, there are no limits related to Tina that will affect you.

> If we were to start with GitHub, what would the migration path to GitLab once support is (hypothetically) added?

This would be simple, you could just migrate from Github to Gitlab and point your project to the new URL (once Gitlab auth is supported)

**Reply by jipiboily** [reply]
Sounds good enough for me, for now! Thanks for the super quick reply.

**Reply by jipiboily** [reply]
Just FYI, we ended up moving to GitHub.

**Comment by genox** [comment]
The team I work with uses a self-hosted GitLab instance because we want sources "closer to home", we only mirror open source to GitHub. While I can technically use GitHub with TinaCMS it creates a dissonance with the common practice. So having the choice to use GitLab would be much appreciated.

On a sidenote: When I got started with tina I was under the impression that it is enough for content to be stored on GitHub (e.g. some `content` folder), while we can use GitLab for the frontend app and use our usual deployment pipelines for infrastructure etc. It would be probably good to make clear somewhere at the start of the onboarding process that tina wants at least the entire `.tina` folder to be present on the linked GH repo.

Thinking out loud, splitting code and content into two repos would also allow for two separate histories that are concerned with either development or content editing, respectively.

I am currently looking into adding a step to our pipeline on GitLab that just pushes the current .tina folder to GitHub and see if that is a viable way to handle it for the time being. But this probably gets a bit hairy once branches diverge while content was written to GH.

**Comment by Fryuni** [comment]
I have not migrated from Forestry to Tina until now exactly because my projects are not hosted on GitHub. (and won't be; GitLab is leagues ahead of GitHub in everything but popularity)

Since Forestry will be turned off in April and what was announced about Tina being Git-based[^1] is still not true (it is _GitHub_-based), I'll have to migrate to another platform. From my research, I'll have to make one for myself since the only one that matched all my needs was Forestry.

For those that want to mix GitLab and GitHub (@genox), GitLab has native support for [Bidirectional mirroring](https://docs.gitlab.com/ee/user/project/repository/mirror/bidirectional.html).

[^1]: [[Ref]](https://forestry.io/blog/forestry.io-end-of-life/) We made this pivot to Tina to innovate with visual editing, Git-backed content. We‘re bringing a drag and drop, visual editing experience to a developer-controlled, open-source, Git-backed CMS

**Reply by scottgallant** [reply]
@Fryuni, if you're interested in building your own connection to GitLab, you can do that with Tina's [self-hosted solution](https://github.com/tinacms/tinacms/discussions/3589). 

You could provide custom onPut/onDelete functions for syncing with a non-github.com repository. If you still want to leverage Tina Cloud for those repos (for auth, user-management, and a hosted data layer), you could do so in combination with your custom, self-hosted Git sync functions.  If you do not want to use Tina Cloud, you could self host it all yourself. 

We're confident this will work but have not done this ourselves yet. So you'd be blazing trail a bit.  

Let us know what you think.

**Comment by voskobovich** [comment]
Hi @scottgallant 

I have the same task.
I want to use the Tina Cloud but store my Tina App in the GitLab repo (because I have a monorepo with shared code parts).

I spend time investigating possible solutions to do this, and my plan is:
1. Create a TinaCMS Cloud account
2. Develop NextJS + TinaCMS app in my monorepo project located in GitLab
3. Connect my app to Tina Cloud by this doc: https://tina.io/docs/tina-cloud/overview/
4. Implement Git Sync functions for GitLab by this doc: https://tina.io/docs/reference/self-hosted/git-provider/make-your-own/
5. Deploy my app to Vercel as usual (not via Tina Dashboard). The Vercel deployment only needs the Tina Admin with a built-in page preview (`admin.my-tina-site.com`). The public website will placed on GitLab Pages.
6. Build the static pages in GitLab Pipelines and publish the website (`my-tina-site.com`).

After these steps, I will have the following:
1. `https://admin.my-tina-site.com` is a dynamic website deployed to Vercel
2. `https://admin.my-tina-site.com/admin/index.html` is the TinaCMS Admin panel with Auth via Tina Cloud
4. `https://my-tina-site.com` is the static website builder in GitLab Pipelines on the commit event
3. User management (accounts, roles, etc.) for my TinaCMS Admin via Tina Cloud

Is my plan correct?

**Reply by scottgallant** [reply]
👋 @voskobovich, what do you want Tina Cloud for? Have you considered just self-hosting Tina? See here https://tina.io/docs/self-hosted/existing-site/

**Reply by voskobovich** [reply]
@scottgallant I am planning to use the Tina Cloud for Auth and User management.

My plan is attempt to implement your suggestion
> If you still want to leverage Tina Cloud for those repos (for auth, user-management, and a hosted data layer), you could do so in combination with your custom, self-hosted Git sync functions.

My project uses PostgreSQL as a database, and I haven't any MongoDB or Redis in my stack to set up a fully self-hosted Tina.

By the way, what does Tina store in this database?

**Reply by scottgallant** [reply]
Is your content stored in PostgreSQL? Do you hope to edit it using TinaCMS? 

Tina stores content in Markdown files (or JSON, YAML, etc) and uses a key value store for its API. The key value store acts like a cache on top of your Markdown content.

Reading the self hosted docs should give you an idea of the best approach. I would also test this (or watch the video) https://tina.io/docs/self-hosted/starters/nextjs-vercel/.

**Reply by voskobovich** [reply]
@scottgallant 

> Is your content stored in PostgreSQL?

No. Currently, I have no content 😄 
I just listed my technologies: GitLab Repo, GitLab Pipelines, GitLab Pages, and PostgreSQL. No one from them doesn't work with TinaCMS.

> Do you hope to edit it using TinaCMS?

No. I need to develop the static page website from scratch and add it to the existing GitLab monorepo.

---
I read the Tina docs. I understand how TinaCMS works and how to use it with Github and Vercel 👌
But my TinaCMS app is placed at GitLab as part of the monorepo project.

For me, using the **built-in** Tina functions and Tina Cloud is the **perfect** solution.

Can I store Tina's content files (markdown, images, etc) in a separate repo from the TinaCMS app?

In the picture below:

- The "TinaCMS App Source codes" are placed on GitLab
- The "TinaCMS Content Files" are placed on GitHub
- The "Website Build Job" combines the source codes and content files and builds static pages

Is this a possible solution?

<img width="676" alt="image" src="https://github.com/tinacms/tinacms/assets/4508703/df4d5a53-7eba-4a59-a09c-d21613d3287a">

**Reply by voskobovich** [reply]
I found in the documentation this

> When working locally, the client is built with the local url (http://localhost:4001/graphql). When in [production mode](https://tina.io/docs/tina-cloud/#prod-mode), clientId, branch and token are used to query Tina Cloud.

and updated my previous picture.

- ~~The "Website Build Job" combines the source codes and content files and builds static pages~~
- The "Website Build Job" fetches content from Tina Cloud via GrapgQL API

<img width="663" alt="image" src="https://github.com/tinacms/tinacms/assets/4508703/1a545d20-c46d-4818-9e17-520fa36697a7">

**Reply by voskobovich** [reply]
WOW!
I just found this guide... 

https://tina.io/guides/tinacms/separate-content-repo/guide/

UPD 1:
Yes, this confirms my guess.
I can store the content in GitHub and the website source codes in GitLab.
For local development, I'll need to clone both to my local machine, and for production, the interaction between the content and the website will be through Tina Cloud.

UPD 2:
But I still need to add images from the content repo to the folder `public/images` to the website root folder.

**Comment by Digizard** [comment]
> Currently we're communicating directly to Github via its REST API, we're in the planning process of moving to a true Git-backed system which will be agnostic. At that point it won't matter where your content comes from. Sorry to say but we don't have a timeline for that yet.

Hi @jeffsee55, is there any update with regards to this after 2 years? I don't see it on the [roadmap](https://tina.io/roadmap/), so is it safe to assume that it is not something deemed important nor coming in the near future?

**Comment by dawidmachon** [comment]
Tina was in plan to be agnostic git. Will it be on roadplan for future or idea dead?

**Comment by ThermScissorpunch** [comment]
Also looking for this integration! We host GitLab ourselves :-)

**Comment by aaron-sac** [comment]
Starting to look into moving from cloud to self-hosted or another CMS solution so that we can use forgejo rather than github given the recent announcement on self-hosted runner prices. I doubt it'll really matter across your whole user base but it might be worth considering that github lock in may now be less acceptable to many folk looking into git backed CMS solutions.
